includes/utilities.inc

```diff
@@ -21,23 +21,59 @@
  *   number. Each an array containing info.
  */
 function islandora_paged_content_get_pages(AbstractObject $object) {
-  $query = <<<EOQ
-PREFIX islandora-rels-ext: <http://islandora.ca/ontology/relsext#>
-SELECT ?pid ?page ?label ?width ?height
-FROM <#ri>
-WHERE {
-  ?pid <fedora-rels-ext:isMemberOf> <info:fedora/{$object->id}> ;
-       <fedora-model:label> ?label ;
-       islandora-rels-ext:isSequenceNumber ?page ;
-       <fedora-model:state> <fedora-model:Active> .
-  OPTIONAL {
-    ?pid <fedora-view:disseminates> ?dss .
-    ?dss <fedora-view:disseminationType> <info:fedora/*/JP2> ;
-         islandora-rels-ext:width ?width ;
-         islandora-rels-ext:height ?height .
- }
+  $pages = array();
+  // Try to use Solr first if enabled, or default to the RI if the Solr query
+  // fails for some reason.
+  if (variable_get('islandora_paged_content_use_solr_for_dimensions', FALSE)) {
+    $pages = islandora_paged_content_get_pages_solr($object);
+    if (empty($pages)) {
+      $pages = islandora_paged_content_get_pages_ri($object);
+    }
+  }
+  else {
+    $pages = islandora_paged_content_get_pages_ri($object);
+  }
+  // Sort the pages into their proper order.
+  $sort = function($a, $b) {
+    $a = (is_array($a) && isset($a['page'])) ? $a['page'] : 0;
+    $b = (is_array($b) && isset($b['page'])) ? $b['page'] : 0;
+    if ($a == $b) {
+      return 0;
+    }
+    return ($a < $b) ? -1 : 1;
+  };
+  uasort($pages, $sort);
+  return $pages;
 }
-ORDER BY ?page
+
+/**
+ * Helper function to retrieve the pages and dimensions from the RI.
+ *
+ * @param AbstractObject $object
+ *   An AbstractObject representing a book.
+ *
+ * @return array
+ *   All the pages in the given paged content object. Ordered by sequence
+ *   number. Each an array containing info.
+ */
+function islandora_paged_content_get_pages_ri($object) {
```

Comment:
@jordandukart

Should be type hinting.

Thanks!

----
William Panting
_Developer_
*[discoverygarden inc.](http://www.discoverygarden.ca) | Managing Digital Content*